Palaso/WritingSystems/LdmlDataMapper.cs

```diff
@@ -493,11 +493,33 @@ public void Write(string filePath, WritingSystemDefinition ws, Stream oldFile, W
 					readerSettings.ProhibitDtd = false;
 					reader = XmlReader.Create(oldFile, readerSettings);
 				}
-				using (var writer = XmlWriter.Create(filePath, CanonicalXmlSettings.CreateXmlWriterSettings()))
+				
+				string backupFilePath = null;
+				if (File.Exists(filePath))
 				{
-					writer.WriteStartDocument();
-					WriteLdml(writer, reader, ws);
-					writer.Close();
+					try
+					{
+						backupFilePath = Path.ChangeExtension(Path.GetTempFileName(), Path.GetExtension(filePath));
```

Comment:
I did intend to delete the backup file at the end and not leave it around, so that was a good catch. I’m not so sure about the merits of using TempFile. For one thing, there is the (very slight) performance hit of creating the file and then overwriting it. In addition, since the backup file might not exist, we can’t just use a using to ensure that it is disposed, so a manual call to Dispose would be required to ensure that it got disposed at the correct time. So I’m leaning toward just calling File.Delete at the appropriate time.

 

From: John Thomson [mailto:notifications@github.com] 
Sent: Thursday, April 30, 2015 3:41 PM
To: sillsdev/libpalaso
Cc: Tom Bogle
Subject: Re: [libpalaso] Prevent overwriting an existing LDML file with an empty one (#219)

 

In Palaso/WritingSystems/LdmlDataMapper.cs <https://github.com/sillsdev/libpalaso/pull/219#discussion_r29462801> :

>                              {
> -                                   writer.WriteStartDocument();
> -                                   WriteLdml(writer, reader, ws);
> -                                   writer.Close();
> +                                   try
> +                                   {
> +                                           backupFilePath = Path.ChangeExtension(Path.GetTempFileName(), Path.GetExtension(filePath));

There's a method on TempFile to get one with a particular extension. Using this would also make sure that the backup gets deleted.

—
Reply to this email directly or view it on GitHub <https://github.com/sillsdev/libpalaso/pull/219/files#r29462801> .  <https://github.com/notifications/beacon/AECJ3iLAENpCEHAIFpHqRKLHj3eR0vyRks5oEnzSgaJpZM4EM3ui.gif>